servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java

```diff
@@ -186,6 +198,80 @@ public final HttpServerBuilder appendConnectionAcceptorFilter(final ConnectionAc
+        if (!OFFLOAD_NONE_STRATEGY.equals(factoryStrategy)) {
+            throw new IllegalArgumentException("Non-offloading filter required offloading " + factoryStrategy);
+        }
+        noOffloadServiceFilters.add(factory);
+        return this;
+    }
+
+    /**
+     * Appends a non-offloading filter to the chain of filters used to decorate the {@link StreamingHttpService} used
+     * by this builder, for every request that passes the provided {@link Predicate}. Filters added via this method
+     * will be executed before offloading occurs and before filters appended via
+     * {@link #appendServiceFilter(StreamingHttpServiceFilterFactory)}.
+     * <p>
+     * Note this method will be used to decorate the {@link StreamingHttpService} passed to
+     * {@link #listenStreaming(StreamingHttpService)} before it is used by the server.
+     * The order of execution of these filters are in order of append, before the filters added with
+     * {@link #appendServiceFilter(StreamingHttpServiceFilterFactory)}. If 3 filters are added as follows:
+     * <pre>
+     *     builder.appendServiceFilter(filter1).appendNonOffloadingServiceFilter(filter2).appendServiceFilter(filter3)
+     * </pre>
+     * accepting a request by a service wrapped by this filter chain, the order of invocation of these filters will be:
+     * <pre>
+     *     filter2 ⇒ filter1 ⇒ filter3 ⇒ service
+     * </pre>
+     *
+     * @param <FF> Super-type of the filter factory which must implement both {@link StreamingHttpServiceFilterFactory}
+     * and {@link HttpExecutionStrategyInfluencer} though non-offloading filters must not influence strategy.
+     * @param predicate the {@link Predicate} to test if the filter must be applied. This must not block.
+     * @param factory {@link StreamingHttpServiceFilterFactory} to append.
+     * @return {@code this}
+     * @throws IllegalArgumentException if the provided filter or predicate requires offloading.
+     */
+    public final <FF extends StreamingHttpServiceFilterFactory & HttpExecutionStrategyInfluencer> HttpServerBuilder
+                appendNonOffloadingServiceFilter(final Predicate<StreamingHttpRequest> predicate,
+                                                 final StreamingHttpServiceFilterFactory factory) {
+        HttpExecutionStrategy predicateStrategy = influenceStrategy(predicate, OFFLOAD_NONE_STRATEGY);
+        if (!OFFLOAD_NONE_STRATEGY.equals(predicateStrategy)) {
+            throw new IllegalArgumentException("Non-offloading predicate required offloading " + predicateStrategy);
+        }
+        HttpExecutionStrategy factoryStrategy = influenceStrategy(factory, OFFLOAD_NONE_STRATEGY);
+        if (!OFFLOAD_NONE_STRATEGY.equals(factoryStrategy)) {
+            throw new IllegalArgumentException("Non-offloading filter required offloading " + factoryStrategy);
+        }
+        appendServiceFilter(service -> new ConditionalHttpServiceFilter(predicate, factory.create(service), service));
```

Comment:
calling `appendServiceFilter` doesn't look quite right. questions:

- does the test verify that no-offloading is done and order is preserved relative to offload filters?
- can you use the code from https://github.com/apple/servicetalk/pull/1710/files#r684634612 (we either need to call `appendNonOffloadingServiceFilter` or use `noOffloadServiceFilters.add`)?

```diff
diff --git a/servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java b/servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
index 8771e3ab0..9323a9243 100644
--- a/servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
+++ b/servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
@@ -215,14 +215,12 @@ public abstract class HttpServerBuilder {
      *     filter2 ⇒ filter1 ⇒ filter3 ⇒ service
      * </pre>
      *
-     * @param <FF> Super-type of the filter factory which must implement both {@link StreamingHttpServiceFilterFactory}
      * and {@link HttpExecutionStrategyInfluencer} though non-offloading filters must not influence strategy.
      * @param factory {@link StreamingHttpServiceFilterFactory} to append.
      * @return {@code this}
      * @throws IllegalArgumentException if the provided filter requires offloading.
      */
-    public final <FF extends StreamingHttpServiceFilterFactory & HttpExecutionStrategyInfluencer> HttpServerBuilder
-                appendNonOffloadingServiceFilter(final FF factory) {
+    public final HttpServerBuilder appendNonOffloadingServiceFilter(final StreamingHttpServiceFilterFactory factory) {
         requireNonNull(factory);
         HttpExecutionStrategy factoryStrategy = influenceStrategy(factory, OFFLOAD_NONE_STRATEGY);
         if (!OFFLOAD_NONE_STRATEGY.equals(factoryStrategy)) {
@@ -250,26 +248,20 @@ public abstract class HttpServerBuilder {
      *     filter2 ⇒ filter1 ⇒ filter3 ⇒ service
      * </pre>
      *
-     * @param <FF> Super-type of the filter factory which must implement both {@link StreamingHttpServiceFilterFactory}
      * and {@link HttpExecutionStrategyInfluencer} though non-offloading filters must not influence strategy.
      * @param predicate the {@link Predicate} to test if the filter must be applied. This must not block.
      * @param factory {@link StreamingHttpServiceFilterFactory} to append.
      * @return {@code this}
      * @throws IllegalArgumentException if the provided filter or predicate requires offloading.
      */
-    public final <FF extends StreamingHttpServiceFilterFactory & HttpExecutionStrategyInfluencer> HttpServerBuilder
-                appendNonOffloadingServiceFilter(final Predicate<StreamingHttpRequest> predicate,
-                                                 final StreamingHttpServiceFilterFactory factory) {
+    public final HttpServerBuilder appendNonOffloadingServiceFilter(final Predicate<StreamingHttpRequest> predicate,
+                                                                    final StreamingHttpServiceFilterFactory factory) {
         HttpExecutionStrategy predicateStrategy = influenceStrategy(predicate, OFFLOAD_NONE_STRATEGY);
         if (!OFFLOAD_NONE_STRATEGY.equals(predicateStrategy)) {
             throw new IllegalArgumentException("Non-offloading predicate required offloading " + predicateStrategy);
         }
-        HttpExecutionStrategy factoryStrategy = influenceStrategy(factory, OFFLOAD_NONE_STRATEGY);
-        if (!OFFLOAD_NONE_STRATEGY.equals(factoryStrategy)) {
-            throw new IllegalArgumentException("Non-offloading filter required offloading " + factoryStrategy);
-        }
-        appendServiceFilter(service -> new ConditionalHttpServiceFilter(predicate, factory.create(service), service));
-        return this;
+        return appendNonOffloadingServiceFilter(service ->
+                new ConditionalHttpServiceFilter(predicate, factory.create(service), service));
     }
 
     /**

```

Reply:
Yes, it was busted good. Now corrected. The test is `HttpServerFilterOrderTest` which infers offloading based upon filter execution order but doesn't definitively check it. (I couldn't figure out how to do that will mocks).